contracts/nft/nft-pre-designed-collection.ral

```diff
@@ -3,31 +3,78 @@
 Contract NFTPreDesignedCollection(
     enumerableNftTemplateId: ByteVec,
     collectionUri: ByteVec,
-    tokenBaseUri: ByteVec,
     collectionOwner: Address,
+    tokenBaseUri: ByteVec,
+    maxSupply: U256,
+    mintPrice: U256,
     mut totalSupply: U256
-) extends NFTCollectionBase(collectionUri, totalSupply) {
-    @using(preapprovedAssets = true, updateFields = true)
-    pub fn mint() -> (ByteVec) {
+) {
+    enum ErrorCodes {
+        IncorrectTokenIndex = 0
+        NFTNotFound = 1
+        TokenOwnerAllowedOnly = 2
+    }
+
+    @using(preapprovedAssets = true, updateFields = true, assetsInContract = true)
+    pub fn mint(index: U256) -> (ByteVec) {
+        checkCaller!(index < maxSupply, ErrorCodes.IncorrectTokenIndex)
+
         let minter = callerAddress!()
-        checkCaller!(minter == collectionOwner, ErrorCodes.TokenOwnerAllowedOnly)
-        let (encodeImmutableFields, encodeMutableFields) = EnumerableNFT.encodeFields!(selfContract!(), totalSupply)
+        let (encodeImmutableFields, encodeMutableFields) = EnumerableNFT.encodeFields!(selfContract!(), index)
 
         let contractId = copyCreateSubContractWithToken!{minter -> ALPH: 1 alph}(
-            toByteVec!(totalSupply),
+            toByteVec!(index),
             enumerableNftTemplateId,
             encodeImmutableFields,
             encodeMutableFields,
             1,
             minter
         )
 
+        transferTokenToSelf!(minter, ALPH, mintPrice)
+
         totalSupply = totalSupply + 1
         return contractId
     }
 
+    @using(checkExternalCaller = false)
+    pub fn nftByIndex(index: U256) -> NFTPreDesignedCollection {
+        assert!(index < maxSupply, ErrorCodes.IncorrectTokenIndex)
+
+        let nftTokenId = subContractId!(toByteVec!(index))
+        assert!(contractExists!(nftTokenId), ErrorCodes.NFTNotFound)
+
+        return NFTPreDesignedCollection(nftTokenId)
```

Comment:
Should it be `INFT(nftTokenId)` here?

Reply:
This method is within the `NFTPreDesignedCollection` contract now, so I thought it makes sense to return `NFTPreDesignedCollection` directly. WDTY?

Reply:
Do you mean the `EnumerableNFT`? Because the `NFTPreDesignedCollection` is a collection, and the `nftTokenId` is an NFT contract id.

And in the latest code, I'm thinking maybe we can reuse the `nftByIndex` function from `NFTCollectionBase`. The only difference between the `nftByIndex` function in the `NFTPublicSaleCollectionRandom` contract and the `nftByIndex` function in the `NFTCollectionBase` is that `NFTPublicSaleCollectionRandom` needs to ensure `index < maxSupply`, while`NFTCollectionBase` needs to ensure `index < totalSupply`. I think we can remove this check, and the reasons are as follows:

1. For the `NFTPublicSaleCollectionRandom` contract, we always ensure that `index < maxSupply` during minting, for the `NFTPublicSaleCollectionSequential` and `NFTOpenCollection` contracts, we mint nfts sequentially, which means the `index < totalSupply` always holds true.
2. In the `nftByIndex` function, we use the `contractExists!` builtin to ensure that the NFT contract indeed exists.

WDYT?

Reply:
> Do you mean the EnumerableNFT? Because the NFTPreDesignedCollection is a collection, and the nftTokenId is an NFT contract id.

Yes, I realized what you mean now. Yes, makes a lot of sense to reuse `nftByIndex` in `NFTCollectionBase`.